feat(WIP): implement llm for metadata conversion

kellytea opened the pull request:
- working on conversion between known formats 
- using litellm
- still need to validate content of the llm response + refactor llm runner configuration

kellytea committed:
feat(WIP): implement llm for metadata conversion

- working on conversion between known formats
-using litellm

sgfost commented on codemeticulous/ai_convert.py:
```diff
@@ -0,0 +1,83 @@
+        "Convert the SOURCE_DATA to match TARGET_MODEL_SCHEMA.\n"
+        "SOURCE_DATA:\n" + json.dumps(source_dict) + "\n\n"
+        "SOURCE_MODEL_SCHEMA:\n" + json.dumps(source_model.schema()) + "\n\n"
+        "TARGET_MODEL_SCHEMA:\n" + json.dumps(target_model.schema()) + "\n\n"
```
I bet context/docs (definitions of terms, etc.) for the schemas might help with results

we could put a `description` for all the fields in the pydantic models so that it would end up in the jsonschema from `schema()`, but this may be a lot of work, even with an initial pass of having AI fill them out from the docs below

alternatively, just throw in the full text of the documentation 

https://codemeta.github.io/terms/
https://datacite-metadata-schema.readthedocs.io/en/4.6/
https://github.com/citation-file-format/citation-file-format/blob/main/schema-guide.md

sgfost commented on codemeticulous/ai_convert.py:
```diff
@@ -0,0 +1,83 @@
+import json
+import os
+
+STANDARDS = {
```
maybe move this to its own file and have convert.py and ai_convert.py both import from it, to avoid defining it twice

sgfost commented on codemeticulous/ai_convert.py:
```diff
@@ -0,0 +1,83 @@
+    messages = prompt_generator(source_instance, source_model, target_model)
+
+    # FIXME: adjust configuration for LiteLLM's standard env var lookup
+    os.environ['OPENROUTER_API_KEY'] = key
```
should be able to just leave this out entirely and assume the user has whatever key they want to use as an env var

kellytea committed:
feat: added llm converion

- need to refactor api key handling
- implement splitting via strategy design
pattern
- add testing

kellytea committed:
fix: simplified llm calls via instructor

- overrides need to set env var

kellytea committed:
fix(WIP): improve "poor" llm model performance

- less capable models like mistral-nemo are failing to map cirtical fields.
- chunking doesn't seem to improve this, might try  refining C-o-T prompt

kellytea committed:
feat: summmarize pydantic schemas with generated descriptions via llm

kellytea committed:
feat:  pruned schemas into llm conversion calls

- removed using instructor because it was having validation errors
- need to fix passing api key in llm call rather than setting env var
- works bi-directionally between codemeticulous and other metedata formats now

sgfost commented on codemeticulous/ai_convert.py:
```diff
@@ -0,0 +1,101 @@
+        source_instance = source_data
+
+    # Create summarized schema of source pydantic model according to data instance
+    source_schema_dict = get_schema_summary(source_model, llm_model, key, source_instance)
```
this should just be cached as a static data file so the descriptions aren't opaquely+non-deterministically generated every time

sgfost commented on codemeticulous/ai_convert.py:
```diff
@@ -0,0 +1,101 @@
+    #     )
+    #     return response
+    try:
+        os.environ["OPENROUTER_API_KEY"] = key #FIXME: avoid setting env var
```
lets remove the key getting passed around and just assume the environment variable is set. A bit more secure and lets us support any api that litellm does with no extra work

kellytea committed:
fix(WIP): consolidate PR fixes

kellytea committed:
fix: refactored schema generation + ai conversion logic

- created command to generate csv schemas all at once
- made generating field types more intuitive than literal
- added guardrail for llm validation fails by allowing retries w/ context of the errors

kellytea committed:
update README.md

kellytea committed:
update README.md

kellytea committed:
test(WIP): first layer evaluation for llm's conversion performance

- test_llm compares between logical conversion & llm conversion
- need to add soft assertion for complex cases, display differences for manual review

kellytea committed:
test(WIP): adding logging by llm model

kellytea committed:
test: refactored observability logs

kellytea committed:
feat: added few shot examples

- cached passed test case llm outputs to use back in few shot prompting in ai_convert
- fixed schema descriptions

kellytea committed:
fix: consolidated copilot suggestions
